Support for private key encryption (KMS) and other bug fixes

aureq opened the pull request:
Here is the main changelog
- apply bucket policy to enforce encryption of private keys
- split IAM managed roles for easier reuse
- reworked of all AWS calls using boto3
- do not guess r53 zone name, get it from configuration 'r53_zone'
- added support for elb custom SSL port via 'elb_port'
- save private keys (account and domain) using AES256 or KMS key via 'kmsKeyArn'
- wait for new IAM server certificate to become available when loading into ELB
- correctly handle missing chain certificate
- a lot more debugs and informative messages
- made function names more consistent
- catching more errors and exceptions whenever possible

aureq committed:
- added kms managed policy
- moved s3 access policy to a separate policy

aureq committed:
added AES encryption class (partial)

aureq committed:
Merge branch 'master' into kms-1

aureq committed:
avoid confusion and explicitly say the elb name is required, not an eventual id

aureq committed:
Merge branch 'master' into kms-1

aureq committed:
removed 'AESCipher' class since encryption will be done via S3-SSE+KMS for now

aureq committed:
updated group policy to cover all encryption operations

aureq committed:
enforce encryption of private keys (*.key.pem) using either kms or S3 managed AES-256

aureq committed:
- cloudformation template support 'AES256' as a generic encryptionn 'key'
- cloudformation template generates a fake kms arn if 'AES256' is provided
- lambda function receives defaullt kms key through 'defaultkey' parameter

aureq committed:
- S3 access policy is broken in 2 entities, one read, one write
- updated lambda role and policy association
- updated policies description

aureq committed:
do not print unecessary exceptions, log only

aureq committed:
checking submitted parameters

aureq committed:
renamed variable 'connection' to 's3_connection', and use 's3_region' variable

aureq committed:
renamed function 'loadAccountKey' to 'load_letsencrypt_account_key'. a bit more info on account key

aureq committed:
changed private key name to '.key.pem' so it matches bucket policy in regards to key encryption (sse)

aureq committed:
renamed function 'createAndSaveKey' to 'create_and_save_key'

aureq committed:
generate and convert key to PEM on-the-fly so 'io' isn't required

aureq committed:
- additional permissions to correctly handle server certs in IAM
- 'UploadServerCertificate' does not allow overwrite, so 'UpdateServerCertificate' is needed
- 'ListServerCertificates' will assist 'UpdateServerCertificate'

aureq committed:
- it's not possible to replace an existing server certificate
  so we need the permission to delete the certificate first
  before uploading it again

aureq committed:
added 'elb_port' and default to 443 if missing

aureq committed:
allow lambda function to describe a load balancer

aureq committed:
replaced boto with boto3 so newer features (kms) are better integrated

aureq committed:
- reworked of all AWS calls using boto3
- do not guess r53 zone name, get it from configuration 'r53_zone'
- added support for elb custom SSL port via 'elb_port'
- save private keys (account and domain) using AES256 or KMS key via 'kmsKeyArn'
- wait for new IAM server certificate to become available when loading into ELB
- correctly handle missing chain certificate
- a lot more debugs and informative messages
- made function names more consistent
- catching more errors and exceptions whenever possible

aureq committed:
Merge remote-tracking branch 'kiddouk/master' into kms-dev

aureq committed:
missing double quote

aureq committed:
added permission 'route53:ListHostedZonesByName' to the lambda function

aureq committed:
removed unecessary permissions

aureq committed:
updated documentation with new parameters

aureq committed:
removed unecessary ';'

kiddouk commented on letslambda.json:
```diff
@@ -58,7 +145,12 @@
+                    {"Ref": "LetsLambdaManagedPolicy"},
+                    {"Ref": "LetsLambdaS3WriteManagedPolicy"},
+                    {"Ref": "LetsLambdaS3ReadManagedPolicy"},
+                    {"Ref": "LetsLambdaKmsKeyManagedPolicy"}
```
I am in doubt if this is the good approach. Generally, we are creating resources in the User's environment, that can be used for many other things.

A cleaner approach would be to have ONE managed policy with multiple statements in order to preserve the split of Actions + Resources.

aureq replied:
Since we save private keys, having a read-only policy allows users to create an ec2 instance role and reuse our pre-defined policies more easily. Typically:
- `LetsLambdaManagedPolicy` is for the lambda execution
- `LetsLambdaS3WriteManagedPolicy` is for lambda but also for anything that would require write access to the bucket (ie: something that would have to write the `yml` file
- `LetsLambdaS3ReadManagedPolicy` is for any AWS resources that would need to retrieve assets in the bucket, but without having the risk of deleting anything
- `LetsLambdaKmsKeyManagedPolicy` to manage access to KMS key access for encryption, but also if private keys have been encrypted with KMS, then this would be required by the relevant consumers (ie: EC2 instances via a role).

kiddouk replied:
SGTM, can we get a doc update on this so the users are aware ?

petemounce replied:
I've found in the past that creating a policy per logical action is reasonable, and then enlisting policies to roles as necessary. That allows a user (with appropriate permission) to assume the role and debug the permissions really easily.

aureq replied:
This is now documented in `kms-dev`. @petemounce you're also part of the contributors/credits.

petemounce replied:
Thanks :)

kiddouk commented on letslambda.py:
```diff
@@ -36,89 +31,75 @@
-    if conf == None:
+
+    try:
+        conf = s3.get_object(Bucket=s3_bucket, Key=letslambda_config)["Body"].read()
```
If I understand well, `boto3` offers a class to manipulate s3 objects in a more natual way : 
http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Object.get

Maybe we could use that ?

aureq replied:
Yes, eventually. the code tried to stay as close as your initial one, but we could the `S3.Object` instead of `S3.Client`

aureq replied:
Not fixing this since both objects provide the exact same feature. I will stick to the high level `S3.Client` class since other part of the code tend to use `Service.Client` class too.

kiddouk commented on letslambda.py:
```diff
@@ -36,89 +31,75 @@
+    newAccountNeeded = False
+    account_key = load_from_s3(conf, 'account.key.pem')
+    if account_key == None:
+        account_key = create_and_save_key(conf, "account.key.pem", conf['kms_key'])
```
Any reason on why are we going from `lowerCamelCase` to `underscore_case` ?

aureq replied:
Well, all functions are `underscore_case`. You will see variations in variable names since it reuses your original code. Could be made more uniform later.

kiddouk replied:
idem. `.rsa` instead of `.pem`

kiddouk commented on letslambda.py:
```diff
@@ -142,127 +348,217 @@ def answer_dns_challenge(client, domain, challenge):
+                s3.put_object(Bucket=conf['s3_bucket'],
+                        Key=s3_key,
+                        Body=content,
+                        ACL='private')
```
The above, is a clear example of kwargs creation  ? can we get that instead ? This will simplify the code.

```
kwargs = {'Bucket': ..., 'Key': ...}
if encrypt == True and kms_key == 'AES256':
  kwargs['ServerSideEncryption'] = "AES256'
...
s3.put_object(**kwargs)
```

kiddouk commented on letslambda.py:
```diff
@@ -1,30 +1,25 @@
+from botocore.config import Config
+from botocore.exceptions import ClientError
+from Crypto import Random
+from Crypto.Cipher import AES
```
unused import.

aureq replied:
used in `conf['region'] = os.environ['AWS_DEFAULT_REGION']`

kiddouk replied:
I am unsure where `AES` is used in the line you are mentioning. Can you provide more details ?

kiddouk commented on letslambda.py:
```diff
@@ -142,127 +348,217 @@ def answer_dns_challenge(client, domain, challenge):
+    except errors.PollError as e:
+        LOG.error("Failed to get certificate issuance for '{0}'.".format(domain['name']))
+        LOG.error("Error: {0}".format(e))
+        print e
```
not required.

kiddouk commented on letslambda.py:
```diff
@@ -142,127 +348,217 @@ def answer_dns_challenge(client, domain, challenge):
+    if 'region' not in event:
+        LOG.error("No bucket region has been provided. Exiting.")
         exit(1)
+    s3_region = event['region']
```
If the region is not specified, I would expect to take the local one. Less config effort.

kiddouk commented:
Overall, this PR is big and contains _A LOT_. I'd prefer something more self contained as we are upgrading libs, introducing new tech and refactoring many things.

Considering that we dont have any unit tests  on the project. This gets tricky.

_BUT_ it looks good! So with a few commits on top of it, I think that it could make it to master in no time.

Keep up the good work :)

aureq committed:
best effort to determine the AWS region for the s3 bucket. fix #1

aureq committed:
made logging more consistent across the whole function. fix #2

aureq committed:
do not print raw exception. fix #3

aureq committed:
removed unused import. fix #4

aureq committed:
use kwargs to improve code readability. fix #5

aureq committed:
- refreshed documentation
- extended documentation on managed policies
- added credit section
- fix #9
- fix #6

petemounce commented on letslambda.json:
```diff
@@ -6,46 +6,133 @@
+                        "Sid": "DenyUnEncryptedPrivKeyUploads",
+                        "Action": ["s3:PutObject"],
+                        "Effect": "Deny",
+                        "Resource": { "Fn::Join" : ["", ["arn:aws:s3:::", { "Ref" : "Bucket" } , "/*.key.pem" ]]},
```
I think it would be a good idea to introduce a parameter `S3BasePath` so that the IAM could be more targetted - this arrangement will affect the whole bucket, which will limit the use of letslambda in accounts that don't want to devote a whole bucket to it (there's a hard upper limit of 100 buckets per account because the names are globally unique).

kiddouk replied:
Yes, but I believe that we dont iterate much much more on this PR to push it to the queue and work on a separate issue for `s3BasePath`

kiddouk replied:
`.rsa` instead of `.pem`

petemounce commented on letslambda.json:
```diff
@@ -6,46 +6,133 @@
+                                "s3:putObject"
+                            ],
+                            "Resource" : [
+                                {"Fn::Join": ["", ["arn:aws:s3:::", {"Ref": "Bucket"}, "/*"]]}
```
As above re: bucket isolation

petemounce commented:
I was going to add that same feedback (as an interested party; I'm not a maintainer!) - I've just discovered this lambda, and it works really well though I fixed a couple typos. I think it might be a reasonable point in its evolution to have some test coverage? I'm not familiar with python, but I can follow the general gist of things here. There's quite a lot going on at this point.

kiddouk commented:
Unit tests on this project will be super tricky I think mainly due to the fact that we make more external calls than real logic in the app.

We could think about some integration tests on a separated AWS account maybe ?

kiddouk commented:
Great ! 

i will give this s spin today and merge back to master if this runs ok on different scenario.

cheers,

petemounce commented:
@kiddouk re: integration tests - what about using https://turq.readthedocs.io/en/latest/ or https://github.com/gabrielfalcao/HTTPretty ? I've not used them, but I've used similar in other languages. So long as the configuration of the AWS clients is accessible, it should be possible to fake away AWS with such tools.

kiddouk commented on letslambda.py:
```diff
@@ -36,89 +30,75 @@
+    LOG.debug("Loading account key from s3")
+
+    newAccountNeeded = False
+    account_key = load_from_s3(conf, 'account.key.pem')
```
This got changed from `.rsa` to `.pem` breaking retro compatibility.

kiddouk commented on letslambda.py:
```diff
@@ -142,127 +344,217 @@ def answer_dns_challenge(client, domain, challenge):
-    return (certificate, key)
+def load_private_key(conf, domain):
+    key = None
+    name = domain['name'] + ".key.pem"
```
`.rsa` versus `.pem`

kiddouk commented:
I dont think that we need to mock the HTTP protocol per say. A simple mocking of responses should be enough I think. I myself use mock2 (for python 2.7).

Mock2 made it to python3 code base.

kiddouk commented:
So I have completely reviewed this and tested it on my account. It works but there are few things that needs to be addressed.
- .rsa extension got changed to .pem. This is a good idea but it breaks with previous installations. this needs to be reverted and CF changed
- The documentation is broken for Debian
- The build.sh script is missing the sourcing of the environment.

I will merge this and created other issues for this that I will fix in `master`
